python/ray/data/_internal/execution/operators/map_operator.py

```diff
@@ -234,53 +245,75 @@ def _add_bundled_input(self, refs: RefBundle):
         """
         raise NotImplementedError
 
-    def _handle_task_submitted(self, task: "_TaskState"):
-        """Handle a newly submitted task, notifying the output queue and updating
-        object store metrics.
-
-        This should be called by subclasses right after a task is submitted.
-
-        Args:
-            task: The task state for the newly submitted task.
-        """
-        # Notify output queue that this task is pending.
-        self._output_queue.notify_pending_task(task)
-
-    @abstractmethod
-    def notify_work_completed(
-        self, ref: Union[ObjectRef[ObjectRefGenerator], ray.ObjectRef]
+    def _submit_data_task(
+        self,
+        gen: StreamingObjectRefGenerator,
+        inputs: RefBundle,
+        task_done_callback: Optional[Callable[[], None]] = None,
     ):
-        """Indicates that a task is done executing OR that a worker is done starting.
-
-        This must be implemented by subclasses.
+        """Submit a new data-handling task."""
+        # TODO(hchen):
+        # 1. Move this to the base PhyscialOperator class.
+        # 2. This method should only take a block-processing function as input,
+        #    instead of a streaming generator. The logic of submitting ray tasks
+        #    can also be capsulated in the base class.
+        task_index = self._next_data_task_idx
+        self._next_data_task_idx += 1
+
+        def output_ready_callback(output: RefBundle):
+            nonlocal task_index
+            # Since output is streamed, it should only contain one block.
+            assert len(output.blocks) == 1
+            for block_ref, _ in output.blocks:
+                trace_allocation(block_ref, "map_operator_output")
+            # Notify output queue that the task has produced an new output.
+            self._output_queue.notify_task_output_ready(task_index, output)
+            # Update object store metrics.
+            allocated = output.size_bytes()
+            self._metrics.alloc += allocated
+            self._metrics.cur += allocated
+            if self._metrics.cur > self._metrics.peak:
+                self._metrics.peak = self._metrics.cur
+
+        def _task_done_callback():
+            nonlocal task_index
+            nonlocal inputs
```

Comment:
I think an alternate pattern to bind these variables correctly is the following:

```
def _task_done(task_index, inputs):
    ...
    
    
DataOpTask(..., lambda task_index=task_index, inputs=inputs: _task_done(task_index, inputs))
```

Reply:
Good suggestion!